src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+
+        break;
+
+      default:
+        break;
+    }
+    stateArray.push({ ...newState });
```

Comment:
Pushing a copy of the object to the array at the end of each loop cycle is correct and follows the best practice of not mutating the original object or arrays. However, consider moving this operation outside of the switch block to avoid repeating it in each case.